client/service/src/lib.rs

```diff
@@ -264,3 +270,174 @@ pub async fn build_relay_chain_interface(
 		)
 	}
 }
+
+/// Parameters given to [`build_network`].
+pub struct BuildNetworkParams<
+	'a,
+	Block: BlockT,
+	Client: ProvideRuntimeApi<Block>
+		+ BlockBackend<Block>
+		+ HeaderMetadata<Block, Error = sp_blockchain::Error>
+		+ HeaderBackend<Block>
+		+ BlockIdTo<Block>
+		+ 'static,
+	RCInterface,
+	IQ,
+> where
+	Client::Api: sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>,
+{
+	pub parachain_config: &'a Configuration,
+	pub client: Arc<Client>,
+	pub transaction_pool: Arc<sc_transaction_pool::FullPool<Block, Client>>,
+	pub para_id: ParaId,
+	pub relay_chain_interface: RCInterface,
+	pub spawn_handle: SpawnTaskHandle,
+	pub import_queue: IQ,
+}
+
+/// Build the network service, the network status sinks and an RPC sender.
+pub async fn build_network<'a, Block, Client, RCInterface, IQ>(
+	BuildNetworkParams {
+		parachain_config,
+		client,
+		transaction_pool,
+		para_id,
+		spawn_handle,
+		relay_chain_interface,
+		import_queue,
+	}: BuildNetworkParams<'a, Block, Client, RCInterface, IQ>,
+) -> sc_service::error::Result<(
+	Arc<NetworkService<Block, Block::Hash>>,
+	TracingUnboundedSender<sc_rpc::system::Request<Block>>,
+	TransactionsHandlerController<Block::Hash>,
+	NetworkStarter,
+)>
+where
+	Block: BlockT,
+	Client: UsageProvider<Block>
+		+ HeaderBackend<Block>
+		+ sp_consensus::block_validation::Chain<Block>
+		+ Send
+		+ Sync
+		+ BlockBackend<Block>
+		+ BlockchainEvents<Block>
+		+ ProvideRuntimeApi<Block>
+		+ HeaderMetadata<Block, Error = sp_blockchain::Error>
+		+ BlockIdTo<Block, Error = sp_blockchain::Error>
+		+ ProofProvider<Block>
+		+ 'static,
+	Client::Api: CollectCollationInfo<Block>
+		+ sp_transaction_pool::runtime_api::TaggedTransactionQueue<Block>,
+	for<'b> &'b Client: BlockImport<Block>,
+	RCInterface: RelayChainInterface + Clone + 'static,
+	IQ: ImportQueue<Block> + 'static,
+{
+	let warp_sync_params = match parachain_config.network.sync_mode {
+		SyncMode::Warp => {
+			let target_block = warp_sync_get::<Block, RCInterface>(
+				para_id,
+				relay_chain_interface.clone(),
+				spawn_handle.clone(),
+			)
+			.await
+			.map_err(|e| format!("Error: {:?}", e))?;
+			Some(WarpSyncParams::WaitForTarget(target_block))
+		},
+		_ => None,
+	};
+
+	let block_announce_validator = BlockAnnounceValidator::new(relay_chain_interface, para_id);
+	let block_announce_validator_builder = move |_| Box::new(block_announce_validator) as Box<_>;
+
+	sc_service::build_network(sc_service::BuildNetworkParams {
+		config: parachain_config,
+		client,
+		transaction_pool,
+		spawn_handle,
+		import_queue,
+		block_announce_validator_builder: Some(Box::new(block_announce_validator_builder)),
+		warp_sync_params,
+	})
+}
+
+/// Creates a new background task to wait for the relay chain to sync up and retrieve the parachain header
+async fn warp_sync_get<B, RCInterface>(
+	para_id: ParaId,
+	relay_chain_interface: RCInterface,
+	spawner: SpawnTaskHandle,
+) -> Result<oneshot::Receiver<<B as BlockT>::Header>, ()>
```

Comment:
We can also remove the result here and return the sender directly. The operation can currently not fail.